engine/resources/hetzner_vm.go

```diff
@@ -0,0 +1,804 @@
+// resource can establish a connection with the API. One Hetzner resource
+// represents one server instance, even if multiple instances are registered
+// under the same project. A resource in the `absent` state only exists as a
+// local blueprint, and does not exist as server instance on Hetzner's side.
+// Serverconfig options can be requested from the Hetzner API, follow
+// https://docs.hetzner.cloud/. All manually registered SSH keys are used when
+// creating a server, and a subset of those can be enabled for rescue mode.
+// TODO: add more flexible selection of SSH keys.
+type HetznerVMRes struct {
+	traits.Base
+	init *engine.Init
+
+	// APIToken specifies the unique API token corresponding to a Hetzner
+	// project. Keep this token private! It provides full access to this
+	// project, so a leaked token will be vulnerable to abuse. Read it from
+	// a local file or the mgmt deploy, or provide it directly as a string.
+	APIToken string `lang:"apitoken"`
+
+	// ServerType determines the machine type as defined by Hetzner.
+	ServerType string `lang:"servertype"`
+	// Datacenter determines where the resource is hosted.
+	Datacenter string `lang:"datacenter"`
+	// Image determines the operating system to be installed.
+	Image string `lang:"image"`
+
+	// State is the desired steady state of the server instance.
+	// Choose between `running`, `off` or `absent`.
+	State string `lang:"state"`
+
+	// WaitInterval is the sleep interval in seconds that is used when waiting
+	// for transient states to converge between intermediate operations.
+	// A zero value causes the waiter to run without delays (burst requests).
+	// Although such burst requests are allowed, it is recommended to use a wait
+	// interval greater than one second for more intensive applications. This
+	// will help to prevent rate limit errors (max. 3600 requests per hour).
+	WaitInterval uint32 `lang:"waitinterval"`
+
+	// WaitTimeout will cancel wait loops if they do not exit cleanly before
+	// the expected time in seconds, in order to detect defective loops and
+	// avoid unnecessary consumption of resources.
+	WaitTimeout uint32 `lang:"waittimeout"`
+
+	// ServerRescueMode specifies the image type used when enabling rescue mode.
+	// The supported image types are `linux32`, `linux64` and `freebsd64`.
+	// Alternatively, use `disabled` to disable rescue mode (default).
+	ServerRescueMode string `lang:"serverrescuemode"`
+
+	// ServerRescueSSHKeys can be used to select a subset of keys that should be
+	// enabled for rescue mode operations over SSH. From all SSH keys known to
+	// the project client, choose a subset of keys by name, as an array of
+	// strings. New keys must first be added manually via the cloud console.
+	// An error is thrown if a given keyname is not recognized by the client.
+	ServerRescueSSHKeys []string `lang:"serverrescuekeys"`
+
+	// AllowDeletion is a server protection policy to prevent unwanted deletion
```

Comment:
discussed in jitsi. tri state.

Reply:
Conclusion:
* Deletion is always allowed when the user explicitly sets `state => "absent"`.
* Deletion for implicit reasons (e.g. rebuilding to switch ServerType) must be protected.
  * `""` throws an error when a deletion is required but the state is not explicitly set to absent.
  * `"ifneeded"` allows implicit deletions if that is required by checkapply to reach all the specified resource fields.
  * `"ignore"` blocks any implicit deletions that might be requested by checkapply, but without error. Continue running.

The same protection should be implemented for poweroffs, to prevent unexpected reboots etc.